lib/PromAlertProxy/Hub.pm

```diff
@@ -0,0 +1,186 @@
+package PromAlertProxy::Hub;
+
+# ABSTRACT: Proxy Prometheus alerts to other places
+
+use 5.028;
+use Moo;
+use experimental qw(signatures);
+
+use Types::Standard qw(CodeRef Object Str Int);
+use Type::Utils qw(class_type role_type);
+use Type::Params qw(compile);
+
+use Prometheus::Tiny;
+use Plack::App::URLMap;
+use Net::Async::HTTP::Server::PSGI;
+use Net::Async::HTTP;
+
+use PromAlertProxy::Logger '$Logger';
+use PromAlertProxy::Endpoint;
+
+has loop => (
+  is       => 'ro',
+  isa      => class_type('IO::Async::Loop'),
+  required => 1,
+);
+
+has http => (
+  is      => 'lazy',
+  isa     => class_type('Net::Async::HTTP'),
+  default => sub ($self) {
+    my $http = Net::Async::HTTP->new;
+    $self->loop->add($http);
+    return $http;
+  },
+);
+
+has _metrics => (
+  is      => 'lazy',
+  isa     => class_type('Prometheus::Tiny'),
+  default => sub ($self) {
+    my $prom = Prometheus::Tiny->new;
+    $prom->declare('promalertproxy_alerts_received_total',
+                   type => 'counter',
+                   help => 'Number of alerts received from Prometheus');
+    $prom->declare('promalertproxy_alerts_dispatched_total',
+                   type => 'counter',
+                   help => 'Number if alerts dispatched to targets');
+    $prom->declare('promalertproxy_alerts_fallback_dispatched_total',
+                   type => 'counter',
+                   help => 'Number if alerts fallback dispatched');
+    $prom;
+  },
+);
+
+has _endpoint => (
+  is      => 'lazy',
+  isa     => class_type('PromAlertProxy::Endpoint'),
+  default => sub ($self) {
+    return PromAlertProxy::Endpoint->new(hub => $self);
+  },
+);
+
+has _app => (
+  is => 'lazy',
+  isa => CodeRef,
+  default => sub ($self) {
+    my $urlmap = Plack::App::URLMap->new;
+
+    # alerts endpoint is the same for both versions
+    $urlmap->map('/api/v1/alerts' => $self->_endpoint->psgi);
+    $urlmap->map('/api/v2/alerts' => $self->_endpoint->psgi);
+
+    $urlmap->map('/metrics' => $self->_metrics->psgi);
+    return $urlmap->to_app;
+  },
+);
+
+sub listen ($self, $ip, $port) {
+  state $check = compile(Object, Str, Int); # XXX IPAddress, Port
+  $check->(@_);
```

Comment:
Yeah, these type checks are even weirder with signatures, because otherwise you don't even need `@_`, but you need to have it for the type checking. (And some of the benefit of the type checking is arity checking, which is already done for you by the signatures.)

Reply:
Yeah, I don't love it but we do what we must.